File: Sources/WickSync/DropboxSyncBackend.swift
import Foundation

/// Dropbox API v2 backend: OAuth 2.0 PKCE sign-in + the four file operations the
/// sync engine needs. No SDK — plain `URLSession` against the documented HTTP API.
///
/// The interactive part of sign-in is injected (`authSession`) by the platform
/// layer (an `ASWebAuthenticationSession` wrapper on macOS/iOS) so this file
/// stays Foundation-only and testable.
@MainActor
public final class DropboxSyncBackend: JournalSyncBackend {
    public static let appKey = "hm5yscsy9a11g0q"
    public static let callbackScheme = "db-hm5yscsy9a11g0q"
    public static let redirectURI = "db-hm5yscsy9a11g0q://2/token"

    private static let authorizeURL = URL(string: "https://www.dropbox.com/oauth2/authorize")!
    private static let tokenURL = URL(string: "https://api.dropboxapi.com/oauth2/token")!
    private static let apiHost = "https://api.dropboxapi.com/2/"
    private static let contentHost = "https://content.dropboxapi.com/2/"

    private let session: URLSession
    private let tokenStore: KeychainTokenStore

    /// Platform hook: opens the system browser at the authorize URL and returns
    /// the callback URL the flow lands on. Arguments: (authorizeURL, callbackScheme).
    public var authSession: ((URL, String) async throws -> URL)?

    public private(set) var accountEmail: String?

    private var accessToken: String?
    private var accessTokenExpiry: Date?
    private var refreshTask: Task<String, Error>?
    /// Loaded lazily from the Keychain on first use.
    private var didProbeKeychain = false
    private var cachedRefreshToken: String?

    public init(
        session: URLSession = {
            let config = URLSessionConfiguration.ephemeral
            config.timeoutIntervalForRequest = 30
            return URLSession(configuration: config)
        }(),
        tokenStore: KeychainTokenStore = KeychainTokenStore(
            service: "com.miaoz.wick.dropbox",
            account: "refresh-token"
        )
    ) {
        self.session = session
        self.tokenStore = tokenStore
    }

    // MARK: - Authorization

    public var isAuthorized: Bool {
        refreshToken() != nil
    }

    public func authorize() async throws -> String {
        guard let authSession else {
            throw SyncBackendError.server(status: 0, message: "auth session not configured")
        }
        let verifier = PKCE.verifier()
        let state = UUID().uuidString

        var components = URLComponents(url: Self.authorizeURL, resolvingAgainstBaseURL: false)!
        components.queryItems = [
            URLQueryItem(name: "client_id", value: Self.appKey),
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "redirect_uri", value: Self.redirectURI),
            URLQueryItem(name: "code_challenge", value: PKCE.challenge(for: verifier)),
            URLQueryItem(name: "code_challenge_method", value: "S256"),
            URLQueryItem(name: "token_access_type", value: "offline"),
            URLQueryItem(name: "state", value: state)
        ]
        let callback = try await authSession(components.url!, Self.callbackScheme)
        guard let callbackComponents = URLComponents(url: callback, resolvingAgainstBaseURL: false),
              callbackComponents.queryItems?.first(where: { $0.name == "state" })?.value == state,
              let code = callbackComponents.queryItems?.first(where: { $0.name == "code" })?.value
        else {
            throw SyncBackendError.authorizationCancelled
        }

        let token = try await tokenRequest(parameters: [
            "grant_type": "authorization_code",
            "code": code,
            "redirect_uri": Self.redirectURI,
            "code_verifier": verifier,
            "client_id": Self.appKey
        ])
        apply(tokenResponse: token)

        let email = (try? await fetchAccountEmail()) ?? ""
        accountEmail = email
        return email
    }

    public func signOut() {
        cachedRefreshToken = nil
        accessToken = nil
        accessTokenExpiry = nil
        accountEmail = nil
        tokenStore.clear()
    }

    private func refreshToken() -> String? {
        if !didProbeKeychain {
            didProbeKeychain = true
            cachedRefreshToken = tokenStore.load()
        }
        return cachedRefreshToken
    }

    private struct TokenResponse {
        var accessToken: String
        var refreshToken: String?
        var expiresIn: TimeInterval
    }

    private func apply(tokenResponse: TokenResponse) {
        accessToken = tokenResponse.accessToken
        // Refresh 5 minutes early to stay clear of mid-flight expiry.
        accessTokenExpiry = Date().addingTimeInterval(max(0, tokenResponse.expiresIn - 300))
        if let refresh = tokenResponse.refreshToken {
            cachedRefreshToken = refresh
            didProbeKeychain = true
            tokenStore.save(refresh)
        }
    }

    private func tokenRequest(parameters: [String: String]) async throws -> TokenResponse {
        var request = URLRequest(url: Self.tokenURL)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        var allowed = CharacterSet.alphanumerics
        allowed.insert(charactersIn: "-._~")
        request.httpBody = parameters
            .map { "\($0.key)=\($0.value.addingPercentEncoding(withAllowedCharacters: allowed) ?? $0.value)" }
            .joined(separator: "&")
            .data(using: .utf8)

        let (data, response) = try await perform(request)
        guard let http = response as? HTTPURLResponse else {
            throw SyncBackendError.transport(message: "no HTTP response")
        }
        guard http.statusCode == 200,
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let access = json["access_token"] as? String
        else {
            // invalid_grant means the code/refresh token is dead.
            let body = String(data: data, encoding: .utf8) ?? ""
            if body.contains("invalid_grant") {
                throw SyncBackendError.needsAuth
            }
            throw SyncBackendError.server(status: http.statusCode, message: body)
        }
        return TokenResponse(
            accessToken: access,
            refreshToken: json["refresh_token"] as? String,
            expiresIn: (json["expires_in"] as? NSNumber)?.doubleValue ?? 4 * 3600
        )
    }

    /// Access token, refreshing via the stored refresh token when needed.
    /// Concurrent callers share one in-flight refresh.
    private func validAccessToken() async throws -> String {
        if let accessToken, let accessTokenExpiry, Date() < accessTokenExpiry {
            return accessToken
        }
        if let refreshTask {
            return try await refreshTask.value
        }
        guard let refresh = refreshToken() else {
            throw SyncBackendError.needsAuth
        }
        let task = Task<String, Error> {
            let response = try await tokenRequest(parameters: [
                "grant_type": "refresh_token",
                "refresh_token": refresh,
                "client_id": Self.appKey
            ])
            apply(tokenResponse: response)
            return response.accessToken
        }
        refreshTask = task
        defer { refreshTask = nil }
        do {
            return try await task.value
        } catch let error as SyncBackendError {
            if error == .needsAuth {
                signOut()
            }
            throw error
        }
    }

    private func fetchAccountEmail() async throws -> String {
        let json = try await rpcCall(path: "users/get_current_account", body: nil)
        return json["email"] as? String ?? ""
    }

    // MARK: - JournalSyncBackend

    public func listChanges(since cursor: String?) async throws -> (entries: [RemoteFileMeta], cursor: String) {
        var collected: [RemoteFileMeta] = []
        var nextCursor: String

        if let cursor {
            nextCursor = cursor
        } else {
            let first = try await rpcCall(path: "files/list_folder", body: [
                "path": "",
                "recursive": true,
                "include_deleted": false,
                "limit": 2000
            ])
            collected.append(contentsOf: parseEntries(first))
            guard let cursor = first["cursor"] as? String else {
                throw SyncBackendError.server(status: 200, message: "list_folder: missing cursor")
            }
            nextCursor = cursor
            var hasMore = first["has_more"] as? Bool ?? false
            while hasMore {
                let page = try await continueListing(nextCursor)
                collected.append(contentsOf: parseEntries(page))
                nextCursor = page["cursor"] as? String ?? nextCursor
                hasMore = page["has_more"] as? Bool ?? false
            }
            return (collected, nextCursor)
        }

        // Incremental: follow continue pagination to the end so the returned
        // cursor always represents "everything up to now".
        while true {
            let page = try await continueListing(nextCursor)
            collected.append(contentsOf: parseEntries(page))
            nextCursor = page["cursor"] as? String ?? nextCursor
            guard page["has_more"] as? Bool ?? false else { break }
        }
        return (collected, nextCursor)
    }

    private func continueListing(_ cursor: String) async throws -> [String: Any] {
        do {
            return try await rpcCall(path: "files/list_folder/continue", body: ["cursor": cursor])
        } catch let error as SyncBackendError {
            // Dropbox reports stale cursors as a 409 'reset' error.
            if case .server(let status, let message) = error,
               status == 409,
               message.contains("reset") || message.contains("expired") {
                throw SyncBackendError.cursorExpired
            }
            throw error
        }
    }

    private func parseEntries(_ json: [String: Any]) -> [RemoteFileMeta] {
        guard let entries = json["entries"] as? [[String: Any]] else { return [] }
        return entries.compactMap { entry in
            let tag = entry[".tag"] as? String
            guard let path = (entry["path_lower"] as? String) ?? (entry["path_display"] as? String)
            else {
                return nil
            }
            switch tag {
            case "file":
                return RemoteFileMeta(
                    path: path,
                    rev: entry["rev"] as? String,
                    contentHash: entry["content_hash"] as? String
                )
            case "deleted":
                return RemoteFileMeta(path: path, rev: nil, contentHash: nil, isDeleted: true)
            default: // folders carry no data for us
                return nil
            }
        }
    }

    public func download(path: String) async throws -> (data: Data, rev: String) {
        let token = try await validAccessToken()
        var request = contentRequest(path: "files/download", token: token)
        request.setValue(
            try apiArg(["path": path]),
            forHTTPHeaderField: "Dropbox-API-Arg"
        )
        let (data, response) = try await perform(request)
        try validate(response: response, data: data, path: path)
        let header = (response as? HTTPURLResponse)?
            .value(forHTTPHeaderField: "Dropbox-API-Result")
        let rev = header
            .flatMap { try? JSONSerialization.jsonObject(with: Data($0.utf8)) as? [String: Any] }
            .flatMap { $0["rev"] as? String } ?? ""
        return (data, rev)
    }

    @discardableResult
    public func upload(path: String, data: Data, ifRev: String?) async throws -> String {
        let token = try await validAccessToken()
        var request = contentRequest(path: "files/upload", token: token)
        let mode: [String: Any] = ifRev.map { [".tag": "update", "update": $0] } ?? [".tag": "add"]
        request.setValue(
            try apiArg(["path": path, "mode": mode, "autorename": false, "mute": true]),
            forHTTPHeaderField: "Dropbox-API-Arg"
        )
        request.httpBody = data
        let (responseData, response) = try await perform(request)
        do {
            try validate(response: response, data: responseData, path: path)
        } catch let error as SyncBackendError {
            if case .server(let status, let message) = error,
               status == 409,
               message.contains("conflict") {
                throw SyncBackendError.writeConflict(path: path)
            }
            throw error
        }
        guard let json = try? JSONSerialization.jsonObject(with: responseData) as? [String: Any],
              let rev = json["rev"] as? String
        else {
            throw SyncBackendError.server(status: 200, message: "upload: missing rev")
        }
        return rev
    }

    public func delete(path: String) async throws {
        do {
            _ = try await rpcCall(path: "files/delete_v2", body: ["path": path])
        } catch let error as SyncBackendError {
            // Deleting something already gone is success for sync purposes.
            if case .server(let status, let message) = error,
               status == 409,
               message.contains("not_found") {
                return
            }
            throw error
        }
    }

    // MARK: - HTTP plumbing

    /// JSON-in/JSON-out call against api.dropboxapi.com. A nil body sends the
    /// literal `null`, which no-argument RPC endpoints expect.
    @discardableResult
    private func rpcCall(path: String, body: [String: Any]?) async throws -> [String: Any] {
        let token = try await validAccessToken()
        var request = URLRequest(url: URL(string: Self.apiHost + path)!)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try body.map { try JSONSerialization.data(withJSONObject: $0) }
            ?? Data("null".utf8)

        let (data, response) = try await perform(request)
        try validate(response: response, data: data, path: path)
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw SyncBackendError.server(status: 200, message: "\(path): unreadable response")
        }
        return json
    }

    private func contentRequest(path: String, token: String) -> URLRequest {
        var request = URLRequest(url: URL(string: Self.contentHost + path)!)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
        return request
    }

    /// Encodes the JSON argument object carried by the `Dropbox-API-Arg` header.
    private func apiArg(_ object: [String: Any]) throws -> String {
        let data = try JSONSerialization.data(withJSONObject: object)
        return String(decoding: data, as: UTF8.self)
    }

    private func perform(_ request: URLRequest) async throws -> (Data, URLResponse) {
        do {
            return try await session.data(for: request)
        } catch let error as URLError {
            throw SyncBackendError.transport(message: error.localizedDescription)
        } catch {
            throw SyncBackendError.transport(message: error.localizedDescription)
        }
    }

    private func validate(response: URLResponse, data: Data, path: String) throws {
        guard let http = response as? HTTPURLResponse else {
            throw SyncBackendError.transport(message: "no HTTP response")
        }
        switch http.statusCode {
        case 200...299:
            return
        case 401:
            // Token rejected outright — refresh didn't help (handled upstream on retry).
            throw SyncBackendError.needsAuth
        case 429:
            let retryAfter = http.value(forHTTPHeaderField: "Retry-After").flatMap(TimeInterval.init)
            throw SyncBackendError.rateLimited(retryAfter: retryAfter)
        default:
            let body = String(data: data, encoding: .utf8) ?? ""
            throw SyncBackendError.server(status: http.statusCode, message: body)
        }
    }
}

File: linux/src/sync/DropboxSyncBackend.cpp
#include "DropboxSyncBackend.h"

#include "JournalModels.h"
#include "PKCE.h"

#include <QEventLoop>
#include <QJsonDocument>
#include <QJsonObject>
#include <QNetworkAccessManager>
#include <QNetworkReply>
#include <QNetworkRequest>
#include <QTimer>
#include <QUrlQuery>

#include <nlohmann/json.hpp>

#include <algorithm>
#include <utility>

namespace wick {
namespace {

using json = nlohmann::json;

QByteArray formEncode(const std::vector<std::pair<std::string, std::string>>& parameters) {
    QByteArray out;
    for (const auto& [k, v] : parameters) {
        if (!out.isEmpty())
            out += '&';
        out += QUrl::toPercentEncoding(QString::fromStdString(k));
        out += '=';
        out += QUrl::toPercentEncoding(QString::fromStdString(v));
    }
    return out;
}

std::string headerValue(QNetworkReply* reply, const QByteArray& name) {
    return QString::fromUtf8(reply->rawHeader(name)).toStdString();
}

} // namespace

DropboxSyncBackend::DropboxSyncBackend(std::shared_ptr<TokenStore> tokenStore,
                                       QNetworkAccessManager* nam)
    : tokenStore_(std::move(tokenStore))
{
    if (nam) {
        nam_ = nam;
    } else {
        ownedNam_ = std::make_unique<QNetworkAccessManager>();
        nam_ = ownedNam_.get();
    }
}

DropboxSyncBackend::~DropboxSyncBackend() = default;

void DropboxSyncBackend::setAccessTokenForTest(std::string token, int expiresInSeconds) {
    accessToken_ = std::move(token);
    const int skewed = std::max(0, expiresInSeconds - 300);
    accessTokenExpiry_ = std::chrono::steady_clock::now() + std::chrono::seconds(skewed);
}

bool DropboxSyncBackend::isAuthorized() const {
    return storedRefreshToken().has_value();
}

std::optional<std::string> DropboxSyncBackend::storedRefreshToken() const {
    if (!didProbeStore_) {
        didProbeStore_ = true;
        if (tokenStore_)
            cachedRefreshToken_ = tokenStore_->load();
    }
    return cachedRefreshToken_;
}

void DropboxSyncBackend::apply(const TokenResponse& token) {
    accessToken_ = token.accessToken;
    const int skewed = static_cast<int>(std::max(0.0, token.expiresIn - 300.0));
    accessTokenExpiry_ = std::chrono::steady_clock::now() + std::chrono::seconds(skewed);
    if (token.refreshToken) {
        if (tokenStore_)
            tokenStore_->save(*token.refreshToken);
        cachedRefreshToken_ = *token.refreshToken;
        didProbeStore_ = true;
    }
}

std::string DropboxSyncBackend::authorize() {
    if (!authSession_) {
        throw SyncBackendError::server(0, "auth session not configured");
    }
    const std::string verifier = PKCE::verifier();
    const std::string state = Uuid::generate().toString();

    QUrl url(endpoints_.authorize);
    QUrlQuery query;
    query.addQueryItem(QStringLiteral("client_id"), QString::fromLatin1(kAppKey));
    query.addQueryItem(QStringLiteral("response_type"), QStringLiteral("code"));
    query.addQueryItem(QStringLiteral("redirect_uri"), QString::fromLatin1(kRedirectURI));
    query.addQueryItem(QStringLiteral("code_challenge"), QString::fromStdString(PKCE::challenge(verifier)));
    query.addQueryItem(QStringLiteral("code_challenge_method"), QStringLiteral("S256"));
    query.addQueryItem(QStringLiteral("token_access_type"), QStringLiteral("offline"));
    query.addQueryItem(QStringLiteral("state"), QString::fromStdString(state));
    url.setQuery(query);

    const QUrl callback = authSession_(url, QString::fromLatin1(kCallbackScheme));
    const QUrlQuery cbQuery(callback);
    const QString gotState = cbQuery.queryItemValue(QStringLiteral("state"));
    const QString code = cbQuery.queryItemValue(QStringLiteral("code"));
    if (gotState.toStdString() != state || code.isEmpty())
        throw SyncBackendError::authorizationCancelled();

    const auto token = tokenRequest({
        {"grant_type", "authorization_code"},
        {"code", code.toStdString()},
        {"redirect_uri", kRedirectURI},
        {"code_verifier", verifier},
        {"client_id", kAppKey},
    });
    apply(token);

    std::string email;
    try {
        email = fetchAccountEmail();
    } catch (...) {
        email.clear();
    }
    accountEmail_ = email;
    return email;
}

void DropboxSyncBackend::signOut() {
    cachedRefreshToken_.reset();
    accessToken_.reset();
    accessTokenExpiry_.reset();
    accountEmail_.reset();
    if (tokenStore_)
        tokenStore_->clear();
}

DropboxSyncBackend::TokenResponse DropboxSyncBackend::tokenRequest(
    const std::vector<std::pair<std::string, std::string>>& parameters) {
    const auto http = perform(endpoints_.token, "POST",
                              {{"Content-Type", "application/x-www-form-urlencoded"}},
                              formEncode(parameters));
    if (http.transportFailed)
        throw SyncBackendError::transport(http.errorString.empty() ? "no HTTP response" : http.errorString);
    if (http.status != 200) {
        if (http.body.find("invalid_grant") != std::string::npos)
            throw SyncBackendError::needsAuth();
        throw SyncBackendError::server(http.status, http.body);
    }
    json j;
    try {
        j = json::parse(http.body);
    } catch (...) {
        throw SyncBackendError::server(http.status, http.body);
    }
    if (!j.contains("access_token") || !j["access_token"].is_string()) {
        if (http.body.find("invalid_grant") != std::string::npos)
            throw SyncBackendError::needsAuth();
        throw SyncBackendError::server(http.status, http.body);
    }
    TokenResponse out;
    out.accessToken = j["access_token"].get<std::string>();
    if (j.contains("refresh_token") && j["refresh_token"].is_string())
        out.refreshToken = j["refresh_token"].get<std::string>();
    if (j.contains("expires_in") && j["expires_in"].is_number())
        out.expiresIn = j["expires_in"].get<double>();
    return out;
}

std::string DropboxSyncBackend::validAccessToken() {
    if (accessToken_ && accessTokenExpiry_ && std::chrono::steady_clock::now() < *accessTokenExpiry_)
        return *accessToken_;
    const auto refresh = storedRefreshToken();
    if (!refresh)
        throw SyncBackendError::needsAuth();
    try {
        const auto response = tokenRequest({
            {"grant_type", "refresh_token"},
            {"refresh_token", *refresh},
            {"client_id", kAppKey},
        });
        apply(response);
        return response.accessToken;
    } catch (const SyncBackendError& e) {
        if (e.kind == SyncBackendError::Kind::needsAuth)
            signOut();
        throw;
    }
}

std::string DropboxSyncBackend::fetchAccountEmail() {
    const auto body = rpcCall("users/get_current_account", std::nullopt);
    try {
        const auto j = json::parse(body);
        if (j.contains("email") && j["email"].is_string())
            return j["email"].get<std::string>();
    } catch (...) {
    }
    return {};
}

std::pair<std::vector<RemoteFileMeta>, std::string>
DropboxSyncBackend::listChanges(const std::optional<std::string>& cursor) {
    std::vector<RemoteFileMeta> collected;
    std::string nextCursor;

    if (!cursor) {
        const json firstBody = {
            {"path", ""},
            {"recursive", true},
            {"include_deleted", false},
            {"limit", 2000},
        };
        const auto firstText = rpcCall("files/list_folder", firstBody.dump());
        json first;
        try {
            first = json::parse(firstText);
        } catch (...) {
            throw SyncBackendError::server(200, "list_folder: unreadable response");
        }
        auto pageMetas = parseEntries(firstText);
        collected.insert(collected.end(), pageMetas.begin(), pageMetas.end());
        if (!first.contains("cursor") || !first["cursor"].is_string())
            throw SyncBackendError::server(200, "list_folder: missing cursor");
        nextCursor = first["cursor"].get<std::string>();
        bool hasMore = first.value("has_more", false);
        while (hasMore) {
            const auto pageText = continueListing(nextCursor);
            json page;
            try {
                page = json::parse(pageText);
            } catch (...) {
                throw SyncBackendError::server(200, "list_folder/continue: unreadable response");
            }
            auto metas = parseEntries(pageText);
            collected.insert(collected.end(), metas.begin(), metas.end());
            if (page.contains("cursor") && page["cursor"].is_string())
                nextCursor = page["cursor"].get<std::string>();
            hasMore = page.value("has_more", false);
        }
        return {collected, nextCursor};
    }

    nextCursor = *cursor;
    while (true) {
        const auto pageText = continueListing(nextCursor);
        json page;
        try {
            page = json::parse(pageText);
        } catch (...) {
            throw SyncBackendError::server(200, "list_folder/continue: unreadable response");
        }
        auto metas = parseEntries(pageText);
        collected.insert(collected.end(), metas.begin(), metas.end());
        if (page.contains("cursor") && page["cursor"].is_string())
            nextCursor = page["cursor"].get<std::string>();
        if (!page.value("has_more", false))
            break;
    }
    return {collected, nextCursor};
}

std::string DropboxSyncBackend::continueListing(const std::string& cursor) {
    const json body = {{"cursor", cursor}};
    try {
        return rpcCall("files/list_folder/continue", body.dump());
    } catch (const SyncBackendError& e) {
        if (e.kind == SyncBackendError::Kind::server && e.status == 409) {
            const std::string& message = e.what();
            if (message.find("reset") != std::string::npos
                || message.find("expired") != std::string::npos) {
                throw SyncBackendError::cursorExpired();
            }
        }
        throw;
    }
}

std::vector<RemoteFileMeta> DropboxSyncBackend::parseEntries(const std::string& jsonText) const {
    json j;
    try {
        j = json::parse(jsonText);
    } catch (...) {
        return {};
    }
    if (!j.contains("entries") || !j["entries"].is_array())
        return {};
    std::vector<RemoteFileMeta> out;
    for (const auto& entry : j["entries"]) {
        if (!entry.is_object())
            continue;
        std::string path;
        if (entry.contains("path_lower") && entry["path_lower"].is_string())
            path = entry["path_lower"].get<std::string>();
        else if (entry.contains("path_display") && entry["path_display"].is_string())
            path = entry["path_display"].get<std::string>();
        else
            continue;
        const std::string tag = entry.value(".tag", "");
        if (tag == "file") {
            RemoteFileMeta m;
            m.path = std::move(path);
            if (entry.contains("rev") && entry["rev"].is_string())
                m.rev = entry["rev"].get<std::string>();
            if (entry.contains("content_hash") && entry["content_hash"].is_string())
                m.contentHash = entry["content_hash"].get<std::string>();
            out.push_back(std::move(m));
        } else if (tag == "deleted") {
            RemoteFileMeta m;
            m.path = std::move(path);
            m.isDeleted = true;
            out.push_back(std::move(m));
        }
    }
    return out;
}

std::pair<std::string, std::string> DropboxSyncBackend::download(const std::string& path) {
    const json arg = {{"path", path}};
    const auto http = performAuthorized([&](const std::string& token) {
        return perform(endpoints_.contentHost + QStringLiteral("files/download"), "POST",
                       {{"Authorization", QByteArray("Bearer ") + QByteArray::fromStdString(token)},
                        {"Content-Type", "application/octet-stream"},
                        {"Dropbox-API-Arg", QByteArray::fromStdString(apiArg(arg.dump()))}},
                       {});
    });
    validate(http, path);
    std::string rev;
    if (!http.dropboxApiResult.empty()) {
        try {
            const auto j = json::parse(http.dropboxApiResult);
            if (j.contains("rev") && j["rev"].is_string())
                rev = j["rev"].get<std::string>();
        } catch (...) {
        }
    }
    return {http.body, rev};
}

std::string DropboxSyncBackend::upload(const std::string& path,
                                       std::string_view data,
                                       const std::optional<std::string>& ifRev) {
    json mode;
    if (ifRev) {
        mode = {{".tag", "update"}, {"update", *ifRev}};
    } else {
        mode = {{".tag", "add"}};
    }
    const json arg = {
        {"path", path},
        {"mode", mode},
        {"autorename", false},
        {"mute", true},
    };
    const auto http = performAuthorized([&](const std::string& token) {
        return perform(endpoints_.contentHost + QStringLiteral("files/upload"), "POST",
                       {{"Authorization", QByteArray("Bearer ") + QByteArray::fromStdString(token)},
                        {"Content-Type", "application/octet-stream"},
                        {"Dropbox-API-Arg", QByteArray::fromStdString(apiArg(arg.dump()))}},
                       QByteArray(data.data(), static_cast<int>(data.size())));
    });
    try {
        validate(http, path);
    } catch (const SyncBackendError& e) {
        if (e.kind == SyncBackendError::Kind::server && e.status == 409) {
            const std::string& message = e.what();
            if (message.find("conflict") != std::string::npos)
                throw SyncBackendError::writeConflict(path);
        }
        throw;
    }
    json j;
    try {
        j = json::parse(http.body);
    } catch (...) {
        throw SyncBackendError::server(200, "upload: missing rev");
    }
    if (!j.contains("rev") || !j["rev"].is_string())
        throw SyncBackendError::server(200, "upload: missing rev");
    return j["rev"].get<std::string>();
}

void DropboxSyncBackend::deletePath(const std::string& path) {
    const json body = {{"path", path}};
    try {
        (void)rpcCall("files/delete_v2", body.dump());
    } catch (const SyncBackendError& e) {
        if (e.kind == SyncBackendError::Kind::server && e.status == 409) {
            const std::string& message = e.what();
            if (message.find("not_found") != std::string::npos)
                return;
        }
        throw;
    }
}

std::string DropboxSyncBackend::rpcCall(const std::string& path, const std::optional<std::string>& jsonBody) {
    const QByteArray body = jsonBody ? QByteArray::fromStdString(*jsonBody) : QByteArray("null");
    const auto http = performAuthorized([&](const std::string& token) {
        return perform(endpoints_.apiHost + QString::fromStdString(path), "POST",
                       {{"Authorization", QByteArray("Bearer ") + QByteArray::fromStdString(token)},
                        {"Content-Type", "application/json"}},
                       body);
    });
    validate(http, path);
    json j;
    try {
        j = json::parse(http.body);
    } catch (...) {
        throw SyncBackendError::server(200, path + ": unreadable response");
    }
    if (!j.is_object())
        throw SyncBackendError::server(200, path + ": unreadable response");
    return http.body;
}

std::string DropboxSyncBackend::apiArg(const std::string& jsonObject) {
    return jsonObject;
}

DropboxSyncBackend::HttpResult DropboxSyncBackend::perform(
    const QString& url,
    const QByteArray& method,
    const std::vector<std::pair<QByteArray, QByteArray>>& headers,
    const QByteArray& body) {
    HttpResult out;
    if (!nam_) {
        out.transportFailed = true;
        out.errorString = "no network manager";
        return out;
    }
    QNetworkRequest req{QUrl(url)};
    req.setTransferTimeout(30000);
    for (const auto& [k, v] : headers)
        req.setRawHeader(k, v);

    QNetworkReply* reply = nam_->sendCustomRequest(req, method, body);
    QEventLoop loop;
    QObject::connect(reply, &QNetworkReply::finished, &loop, &QEventLoop::quit);
    loop.exec();

    if (reply->error() == QNetworkReply::TimeoutError
        || reply->error() == QNetworkReply::ConnectionRefusedError
        || reply->error() == QNetworkReply::HostNotFoundError
        || reply->error() == QNetworkReply::SslHandshakeFailedError
        || reply->error() == QNetworkReply::UnknownNetworkError
        || reply->error() == QNetworkReply::TemporaryNetworkFailureError
        || reply->error() == QNetworkReply::NetworkSessionFailedError
        || reply->error() == QNetworkReply::RemoteHostClosedError
        || reply->error() == QNetworkReply::OperationCanceledError) {
        // HTTP status errors (401/409/429) also set QNetworkReply::error, so only
        // treat true transport failures when there is no HTTP status.
        const QVariant statusVar = reply->attribute(QNetworkRequest::HttpStatusCodeAttribute);
        if (!statusVar.isValid()) {
            out.transportFailed = true;
            out.errorString = reply->errorString().toStdString();
            reply->deleteLater();
            return out;
        }
    }

    out.status = reply->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
    out.body = reply->readAll().toStdString();
    out.dropboxApiResult = headerValue(reply, "Dropbox-API-Result");
    const QByteArray retry = reply->rawHeader("Retry-After");
    if (!retry.isEmpty()) {
        bool ok = false;
        const double seconds = QString::fromUtf8(retry).toDouble(&ok);
        if (ok)
            out.retryAfter = seconds;
    }
    if (out.status == 0 && reply->error() != QNetworkReply::NoError) {
        out.transportFailed = true;
        out.errorString = reply->errorString().toStdString();
    }
    reply->deleteLater();
    return out;
}

DropboxSyncBackend::HttpResult DropboxSyncBackend::performAuthorized(
    const std::function<HttpResult(const std::string& token)>& make) {
    std::string token = validAccessToken();
    bool didRefresh = false;
    while (true) {
        HttpResult http = make(token);
        if (http.transportFailed)
            throw SyncBackendError::transport(http.errorString.empty() ? "no HTTP response" : http.errorString);
        if (http.status != 401)
            return http;
        if (didRefresh)
            throw SyncBackendError::needsAuth();
        accessToken_.reset();
        accessTokenExpiry_.reset();
        token = validAccessToken();
        didRefresh = true;
    }
}

void DropboxSyncBackend::validate(const HttpResult& http, const std::string& path) const {
    (void)path;
    if (http.transportFailed)
        throw SyncBackendError::transport(http.errorString.empty() ? "no HTTP response" : http.errorString);
    if (http.status >= 200 && http.status <= 299)
        return;
    if (http.status == 401)
        throw SyncBackendError::needsAuth();
    if (http.status == 429)
        throw SyncBackendError::rateLimited(http.retryAfter);
    throw SyncBackendError::server(http.status, http.body);
}

} // namespace wick
